curvefs/src/mds/fs_manager.cpp

```diff
@@ -42,39 +43,118 @@ bool FsManager::Init() {
+        sleeper_.interrupt();
+        backEndThread_.join();
+        LOG(INFO) << "stop FsManager ok.";
+    } else {
+        LOG(INFO) << "FsManager not running.";
+    }
+
     fsStorage_->Uninit();
     spaceClient_->Uninit();
+    LOG(INFO) << "FsManager Uninit ok.";
 }
 
-FSStatusCode FsManager::CreateFs(const std::string& fsName,
-                                 FSType fsType, uint64_t blockSize,
-                                 const FsDetail& detail, FsInfo* fsInfo) {
+void FsManager::ScanFs(const FsInfoWrapper& wrapper) {
+    if (wrapper.GetStatus() != FsStatus::DELETING) {
+        return;
+    }
+
+    std::list<PartitionInfo> partitionList;
+    topoManager_->ListPartitionOfFs(wrapper.GetFsId(), &partitionList);
+    if (partitionList.size() == 0) {
+        LOG(INFO) << "fs has no partition, delete fs record, fsName = "
+                  << wrapper.GetFsName() << ", fsId = " << wrapper.GetFsId();
+        FSStatusCode ret = fsStorage_->Delete(wrapper.GetFsName());
+        if (ret != FSStatusCode::OK) {
+            LOG(ERROR) << "delete fs record fail, fsName = "
+                       << wrapper.GetFsName()
+                       << ", errCode = " << FSStatusCode_Name(ret);
+        }
+        return;
+    }
+
+    for (const PartitionInfo& partition : partitionList) {
+        if (partition.status() != PartitionStatus::DELETING) {
+            LOG(INFO) << "delete fs partition, fsName = " << wrapper.GetFsName()
+                      << ", partitionId = " << partition.partitionid();
+            // send rpc to metaserver, get copyset members
+            std::set<std::string> addrs;
+            if (TopoStatusCode::TOPO_OK !=
+                topoManager_->GetCopysetMembers(
+                    partition.poolid(), partition.copysetid(), &addrs)) {
+                LOG(ERROR) << "delete partition fail, get copyset "
+                              "members fail"
+                           << ", poolId = " << partition.poolid()
+                           << ", copysetId = " << partition.copysetid();
+                return;
+            }
+
+            FSStatusCode ret = metaserverClient_->DeletePartition(
+                partition.poolid(), partition.copysetid(),
+                partition.partitionid(), addrs);
+            if (ret != FSStatusCode::OK &&
+                ret != FSStatusCode::UNDER_DELETING) {
+                LOG(ERROR) << "delete partition fail, fsName = "
+                           << wrapper.GetFsName()
+                           << ", partitionId = " << partition.partitionid()
+                           << ", errCode = " << FSStatusCode_Name(ret);
+            }
+        }
+    }
+}
+
+void FsManager::BackEndFunc() {
+    while (sleeper_.wait_for(
+        std::chrono::seconds(backEndThreadRunInterSec_))) {
+        std::vector<FsInfoWrapper> wrapperVec;
+        fsStorage_->GetAll(&wrapperVec);
+        for (const FsInfoWrapper& wrapper : wrapperVec) {
+            ScanFs(wrapper);
+        }
```

Comment:
this manner is also differing from your original design, in your design, you say when deleting fs just start one backend delete task, and when MDS restarting, regenerate delete tasks. but your implementation is looping all existing fs.

Reply:
looping all existing fs is simpler to implement